draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -816,327 +1107,323 @@ Input: alpha, an octet string to be hashed.
+Z^2 * u^4 + Z * u^2 == 0. This includes u == 0, and may include
+other values depending on Z. Implementations must detect
+this case and set x1 = B / (Z * A), which guarantees that g(x1)
+is square by the condition on Z given above.
 
 Operations:
 
 ~~~
-1.    u = hash2base(alpha || I2OSP(0,1))
-2.   x1 = hash2base(alpha || I2OSP(1,1))
-3.  gx1 = x1^3 + A*x1 + B
-4.   x2 = (-B / A) * (1 + 1 / (u^4 * gx1^2 + u^2 * gx1))
-5.  gx2 = x2^3 + A*x2 + B
-6.   x3 = u^2 * gx1 * x2
-7.  gx3 = x3^3 + A*x3 + B
+1.   u = hash2base(alpha)
+2. den = inv0(Z^2 * u^4 + Z * u^2)
+3.  x1 = (-B / A) * (1 + den)
+4.  If den == 0, set x1 = B / (Z * A)
+5. gx1 = x1^3 + A * x1 + B
+6.  x2 = Z * u^2 * x1
+7. gx2 = x2^3 + A * x2 + B
 8.  If gx1 is square, set x = x1 and y = sqrt(gx1)
 9.  If gx2 is square, set x = x2 and y = sqrt(gx2)
-10. If gx3 is square, set x = x3 and y = sqrt(gx3)
-11. Output h*(x, y)
+10. If sgn0(u) != sgn0(y), set y = -y
+11. Output h * (x, y)
 ~~~
 
 #### Implementation
 
-The following procedure implements SWU's algorithm in a straight-line fashion.
+The following procedure implements the simplified SWU algorithm in a
+straight-line fashion. This implementation is optimized for the case
+that q = 3 (mod 4), which applies to P-256.
+For discussion of how to generalize to q = 1 (mod 4), see
+{{WB19}} (Section 4) or the example code found at {{github-repo}}.
 
 ~~~
-map2curve_swu(alpha)
-
+map2curve_simple_swu(alpha)
 Input: alpha, an octet string to be hashed.
 Output: (x, y), a point on E.
 
+Precondition: q = 3 (mod 4)
+
 Constants:
-1. c1 = - B / A
+1.  c1 = -B / A
+3.  c2 = -1 / Z
+4.  c3 = sqrt(-Z^3)
 
 Steps:
-1.    u = hash2base(alpha || I2OSP(0,1))
-2.   x1 = hash2base(alpha || I2OSP(1,1))
-3.  gx1 = x1^2
-4.  gx1 = gx1 + A
-5.  gx1 = gx1 * x1
-6.  gx1 = gx1 + B       // gx1 = x1^3 + A*x1 + B
-7.   t1 = u^2
-8.   t1 = t1 * gx1      // t1 = u^2 * gx1
-9.   t2 = t1^2
-10.  t2 = t2 + t1      
-11.  t2 = 1 / t2        // t2 = 1 / (t1^2 + t1)
-12.  x2 = t2 + 1
-13.  x2 = x2 * c1       // x2 = (-B/A) * (1 + 1/(u^4*gx1^2 + u^2*gx1))
-14. gx2 = x2^2
-15. gx2 = gx2 + A
-16. gx2 = gx2 * x2
-17. gx2 = gx2 + B       // gx2 = x2^3 + A*x2 + B
-18.  x3 = t1 * x2       // x3 = u^2 * gx1 * x2
-19. gx3 = x3^2
-20. gx3 = gx3 + A
-21. gx3 = gx3 * x3
-22. gx3 = gx3 + B       // gx3 = x3^3 + A*x3 + B
-23.  e1 = is_square(gx1, q)
-24.  e2 = is_square(gx2, q)
-25.   x = CMOV(x3, x2, e2)    // If e2=True, x = x2, else x = x3
-26.   x = CMOV(x, x1, e1)     // If e1=True, x = x1, else x = x
-27.  gx = CMOV(gx3, gx2, e2)  // If e2=True, gx = gx2, else gx = gx3
-28.  gx = CMOV(gx, gx1, e1)   // If e1=True, gx = gx1, else gx = gx
-29.   y = sqrt(gx, q)
-30. Output h*(x, y)
+1.    u = hash2base(alpha)
+2.   t1 = Z * u^2
+3.   t2 = t1^2
+4.   x1 = t1 + t2
+5.   x1 = inv0(x1)
+6.   e1 = x1 == 0
+7.   x1 = x1 + 1
+8.   x1 = CMOV(x1, c2, e1)   // if (t1 + t2) == 0, set x1 = -1 / Z
+7.   x1 = x1 * c1      // x1 = (-B / A) * (1 + (1 / (Z^2 * u^4 + Z * u^2)))
```

Comment:
nit: comment is not aligned with the others. Was that due to column overflow?

Reply:
Yes, was trying to prevent the HTML version from needing a horizontal scrollbar for this code snippet. Which style do you prefer?

Reply:
Lack of overflow is best, so let's keep it as is.